File: two_pointers/2302.py
#  Tc = O(2 * n) => O(n) [each element visited twice using i once, j once]
#  Tc = O(1)

class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        n = len(nums)
        result = 0
        sum = 0
        i = 0
        j = 0
        while j < n:
            sum += nums[j]
            while i <= j and sum * (j - i + 1) >= k:
                sum -= nums[i]
                i += 1
            result += j - i + 1
            j += 1
        return result

File: two_pointers/2302.cpp
// Tc = O(2 * n) => O(n) [each element visited twice using i once, j once]
// Tc = O(1)

class Solution {
public:
    long long countSubarrays(vector<int>& nums, long long k) {
        int n = nums.size();
        long long result = 0;
        long long sum = 0;
        int i = 0;
        int j = 0;
        while (j < n) {
            sum += nums[j];
            while (i <= j && sum * (j - i + 1) >= k) {
                sum -= nums[i]; //shrink the window
                i++;
            }
            result += (j - i + 1);
            j++;
        }
    return result;
    }
};